### DISCUSSION TITLE
Crash course on lambda calculus

### DISCUSSION THREAD

**Discussion by josePereiro** [josePereiro/TankeFactory#5] [Discussions]
### 1. The lambda calculus
Alonzo Church came up with a fairly nice and simple calculus that can be used as a basis for both computation and mathematics. In his version, there are three types of symbolic expressions:
1.  $x$ if $x$ is a a symbol (variable), where we presume we have an infinite set of symbols to draw from.
2. $(x \to M)$ if $x$ is a symbol and $M$ is a lambda calculus term (abstraction)
3. $(M~N)$ if $M$ and $N$ are lambda calculus terms (application),

Ok, I'm not getting 2 and 3. It says "This sort of abstractly models the idea of functions and applying functions to values, but with the oddity that everything is a function. The “semantics” are defined by an equivalence relation on lambda calculus terms". I get the part that all information is encoded in the relationships between functions, but, how it is translated from the syntax in 2 and 3?
Is it 2 just a function from domain $x$ to image $M$? Why is $M$ capitalized? Is it 3 just a compound functions? What is the rationale of the "names" abstraction (2) and applications (3)?

**Comment by ta-rekt** [comment]
Yes, abstraction just defines an unnamed function taking a variable x and mapping it to a term t, so it means exactly the same thing as f : x -› t, but here the function has a name "f". Also he is using non-standard notation. The more conventional way of expressing abstraction in lambda calculus is this: λx.t (which is why it's called λ-calculus). M is capitalized for no specific reason. it could be any symbol denoting a term.

The general idea of lambda calculus is similar to having a set of Legos that can fit together in different ways so that you can build more complicated things from a few basic pieces. A single lego piece is a "variable". A "term" is just any object that it's possible to build using legos (including variables themselves). So everything expressible in lambda calculus is called a term. Terms can be constructed from simple variables, or from other terms, in two different ways: application and abstraction (by analogy: two different ways that lego objects can be combined together to make a bigger object).

if you have two terms t1 and t2, you can combine them by creating a function λt1.t2, which is a new term. t1 is not necessarily a variable (an "atomic" term) for this term to be well-formed.

you can also combine them to create an application of a function, written t1 t2, which is exactly the same as t1(t2) in traditional function notation. Here of course there is a constraint that t1 should be a function, meaning an abstraction term, or else the term t1 t2 is not well-formed.

the notation M[x -> y] or more conventionally M[y/x] just means take the term M and replace all instances of variable x in M by variable y. now you can see that (λx.t1) t2, meaning "apply function λx.t1 with input t2" is equivalent to t1[x -> t2]. This is called β-reduction, which you can denote by β-> for short.

Two more important things: 
1. you can partially apply a function with n variables e.g. give it only one variable, and you get as output another function with n-1 variables, and the leftmost variable in the input will disappear. So the application (λxy.t) z β-reduces to λy.t[x -> z]. 
2. application is left-associative and abstraction is right-associative: t1 t2 t3 means (t1 t2) t3, and λx.λy z means λx.(λy z)

Church proved that any computable function can be expressed in λ-calculus, and it is an equivalent formulation of universal computation to Turing machines. So out of these simple rules, you can build any lego machine you want, no matter how complex. 

For example, you can encode boolean logic like this:

True := λxy.x
False := λxy.y
if f then x else y := λfxy.((f x) y) (or simply λfxy.f x y)
x And y := if x then y else False
x Or y := if x then True else y
Not x := if x then False else True

try "if" with f = True:
(if f then x else y) True x y
:= (λfxy.f x y) True x y
β-> (λxy.True x y) x y
β-> (λxy.x) x y
β-> x

You can also encode natural numbers and arithmetic:

**Natural numbers**
0 := λfx.x
Succ n := λnfx.n f (f x) (take a natural n and increment it by one by applying "f" once)
1 := Succ 0 =: (λnfx.n f (f x)) (λfx.x) 
β-> λfx.(λfx.x) f (f x)
β-> λfx.f x

2 := Succ 1 β-> λfx.f (f x)
3 := Succ 2 β-> λfx.f (f (f x))
etc.

**Addition**
n + m := λnmfx.n f (m f x) (apply function "f" m times to x, then pass the result to n to apply it n more times -> n + m total applications)
1 + 2 := (λnmfx.n f (m f x)) (λfx.f x) (λfx.f (f x))
β-> λfx.(λfx.f x) f ((λfx.f (f x)) f x)     (replaced n and m by 1 and 2)
β-> λfx.(λfx.f x) f (f (f x))
β-> λfx.f (f (f x))
:= 3

**Multiplication**
n * m := λnmfx.n (m f) x (take the function which applies "f" m times and apply **it** n times to x, resulting in n * m total applications)
2 * 2 := try it!

**Reply by ta-rekt** [reply]
By the way I have excellent and very detailed slides from my masters if you want more background

**Reply by josePereiro** [reply]
Send them to me, but I will attack the problem with several sources for better results. I'm starting from scratch here. Understanding your post above is the next target!
